src/DittoMachine.sol

```diff
@@ -419,9 +421,68 @@ contract DittoMachine is ERC721, ERC721TokenReceiver {
             from,
             cloneShape.worth + subsidy
         );
+    }
+
+    /**
+     * @dev will allow NFT sellers to sell by safeTransferFrom-ing directly to this contract.
+     * @param data will contain ERC20 address that the seller wishes to sell for
+     * allows specifying selling for the floor price
+     * @return returns received selector
+     */
+    function onERC721Received(
+        address,
+        address from,
+        uint256 id,
+        bytes calldata data
+    ) external returns (bytes4) {
+        (address ERC20Contract, bool floor) = abi.decode(data, (address, bool));
+
+        onTokenReceived(from, msg.sender /*ERC721 contract address*/, id, ERC20Contract, floor, false);
+
         return this.onERC721Received.selector;
     }
 
+    function onERC1155Received(
+        address,
+        address from,
+        uint256 id,
+        uint256 amount,
+        bytes calldata data
+    ) external returns (bytes4) {
+        if (amount != 1) {
+            revert AmountInvalid();
+        }
+        // address ERC1155Contract = msg.sender;
+        (address ERC20Contract, bool floor) = abi.decode(data, (address, bool));
+
+        onTokenReceived(from, msg.sender /*ERC1155 contract address*/, id, ERC20Contract, floor, true);
+
+        return this.onERC1155Received.selector;
+    }
+
+    function onERC1155BatchReceived(
+        address,
+        address from,
+        uint256[] calldata ids,
+        uint256[] calldata amounts,
+        bytes calldata data
+    ) external returns (bytes4) {
+        for (uint256 i; i < amounts.length; i++) {
+            if (amounts[i] != 1) {
+                revert AmountInvalid();
+            }
+        }
+        // address[] memory ERC20Contracts = new address[](ids.length);
+        // bool[] memory floors = new bool[](ids.length);
+        (address[] memory ERC20Contracts, bool[] memory floors) = abi.decode(data, (address[], bool[]));
```

Comment:
The 1155 batch transfer test is failing here. I think `abi.decode` is having issues with dynamically sized arrays. If we stay with a single erc20 contract and bool then the test will pass, but I would prefer if we could pass arbitrary details for each token in the batch transfer.

Reply:
Will this work? We can slice `data` for `address[]` and `bool[]` based on ids.length, and then decode them separately.

Reply:
tried this:
```solidity
address[] memory ERC20Contracts = abi.decode(data[:ids.length*32], (address[]));
bool[] memory floors = abi.decode(data[ids.length*32:], (bool[]));
```
still getting `revert()` error on the test.